<p><img src="https://files.cdn.thinkific.com/file_uploads/133292/images/a9e/c1a/bc3/P02.jpg"></p>

`MCSDSC02P01V3______`

`MCSDSC02P01V2______`

`MCSDSELIMC02P01V1010621`

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Привет, Анна! Ты первопроходка, да еще и с самой лучшей итерацией на текущий момент. Критических замечаний мало, но есть доработки.

Как обычно, заголовки (комментариев) будут трех цветов. <font color="89DD00">Зеленый</font> — все правильно (и если комментария нет, то это также значит, что все правильно). <font color="FFB800">Желтый</font> — можно доработать. <font color="FF0F00">Красный</font> — нужно исправить.

Постарайся учесть эти комментарии в ходе доработки проекта и не удаляй их.

## Постановка задачи

Изучить данные сервиса с тестового запуска приложения 'Без очереди', которое позволяет бронировать столик в партнерском ресторане на какое-то конкретное время и определнное количество человек, а также делать заказ заранее, чтобы не ждать официанта и счет. 

Выяснить:

* На какие кафе и на какие дни приходится больше всего заказов;
* В какой временной промежуток приложением пользуются чаще: дневной или вечерний;
* Насколько большими компаниями  (или наоборот поодиночке) приходят в кафе.


<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Да, описание сервиса и задача от менеджера, все четко.

## Импорт данных и библиотек

In [1]:
import pandas as pd

Прочитаем файл `bez_ocheredi.csv` и сохраним его в переменную `bo_data`.

In [2]:
#bo_data = pd.read_csv('/Users/ana/Documents/mcs_ds_02/data/bez_ocheredi.csv')

bo_data = pd.read_csv('/Users/elimilo/Desktop/mcs_ds_02/data/bez_ocheredi.csv')

Получим первые 10 строк таблицы.

In [138]:
bo_data.head(10)

userID    ресторан     weekday  ordered    Столик для 
0  b912c3c6a6ad   Iskra Bar     Monday     10:00             1
1  2dac0150717d   Iskra Bar     Monday     13:30             2
2  ba83aefab5dc   Iskra Bar    Tuesday     12:30             3
3  02996f1dd2ea   Iskra Bar  Wednesday     12:00             2
4  7c90be56d2ab   Iskra Bar   Thursday     14:30             3
5  51e0962387f7      Там Ем     Friday     12:30             2
6  c4f260a2b48a         NaN   Saturday     15:00             6
7  6b71309d6a8a  Крафтейнер     Monday     10:30             1
8  c5b969cb63a2  Бар Сибирь  Wednesday     15:30             2
9  b1570962e68c  Бар Сибирь  Wednesday     12:30             3

Получим общую информацию о данных из таблицы.

In [139]:
bo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1024 entries, 0 to 1023
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userID        1024 non-null   object
 1    ресторан     959 non-null    object
 2    weekday      1024 non-null   object
 3    ordered      1024 non-null   object
 4    Столик для   1024 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 40.1+ KB


Столбцы в `bo_data` содержат следующую информацию:

* ID пользователя приложения; 
* название ресторана, в котором забронирован столик;
* день недели и время брони;
* количество людей за столиком.

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Здесь через запятую не очень, используй буллиты:

* userID — ID пользователя
* ...

### Выводы

* Размер датасета 5 столбцов на 1024 строки.
* Уникальных записей 1000, что говорит о наличии полных дубликатов в датасете.
* В колонке 'ресторан' имеются пропуски;
* В датасете четыре колонки с типом данных object ('userID', 'ресторан', 'weekday', 'ordered') и одна с типом int64 ('Столик для'). 

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Да, нужно указать на структуру данных, а также наличие дубликатов и пропусков. На счет последнего — посмотри на колонку "ресторан", там non-null меньше.

Также добавлю, что атрибут `dtypes` поможет вывести только типы данных.

## Предобработка данных

Получим названия столбцов.

In [140]:
print(bo_data.columns)

Index(['userID ', ' ресторан ', ' weekday', ' ordered ', ' Столик для '], dtype='object')


Обращаем внимание, что помимо названий на разном языке, в них есть пробелы, которые затрудняют доступ к данным. Переименуем столбцы и проверим результат.

In [141]:
bo_data.set_axis(['user_ID', 'restaurant', 'weekday', 'ordered_time', 'table_for'], axis='columns', inplace=True)

In [142]:
print(bo_data.columns)

Index(['user_ID', 'restaurant', 'weekday', 'ordered_time', 'table_for'], dtype='object')


<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Так мы именно переименовываем столбцы, а не создаем новый набор названий.

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Не надо переименовывать столбцы через атрибут columns, потому что так ты неявно учитываешь порядок столбцов. Если порядок поменяется, то код не будет работать. 

Все столбцы можно переименовать с помощью одного метода — `set_axis()`. Передавай ему список с названиями столбцов, `axis='columns'`, `inplace = True` (последний аргумент — переименовывает именно существующий датафрейм, а не создает копию). Если хочешь переименовать только некоторые, то `rename()` тоже подойдет, но опять же указывай `inplace = True`.

Еще момент: userID — это camel case, а table_for — snake_case. Лучше на чем-то одном остановиться, а также ordered назвать очевиднее: например, ordered_time.

Еще раз проверим данные на наличие пропусков, просуммировав их.

In [143]:
bo_data.isnull().sum()

user_ID          0
restaurant      65
weekday          0
ordered_time     0
table_for        0
dtype: int64

Менеджер проекта предупреждал, что некоторые заказы заводились на тестовую карточку кафе без названия, поэтому есть смысл заполнить такой пропуск словом admin.

In [144]:
bo_data['restaurant'].fillna('admin', inplace=True)

Проверяем, что пропусков не осталось.

In [145]:
bo_data.isnull().sum()

user_ID         0
restaurant      0
weekday         0
ordered_time    0
table_for       0
dtype: int64

Установим количество явных дубликатов. Если найдутся, то удалим и снова проверим, что их не осталось.

In [146]:
bo_data.duplicated().sum()


24

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Любопытно, но для этой задачи в pandas лучше использовать цепочку методов `duplicated().sum()`.

In [3]:
bo_data.drop_duplicates(keep="first", inplace=True)

In [4]:
bo_data.duplicated().sum()

0

In [5]:
bo_data.loc[20:27]

userID   ресторан     weekday  ordered    Столик для 
20  ea12aec32ec3    Северия    Tuesday     12:30             4
21  838a82cec0a6        NaN  Wednesday     13:00             4
26  de310e6aae86    Северия   Thursday     12:00             2
27  08b24e170109    Северия   Saturday     11:00             3

<h3><font color="#FF0F00">Комментарий ассистента 2</font></h3>

Здесь не увидел изменений :-))

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Вывел срез, в котором были дубликаты, чтобы показать, что если не использовать метод `reset_index()` после `drop_duplicates()`, то получается дыра (21-26). Ему также нужно передать `drop=True`, чтобы он создавал не отдельный столбец, а обновлял именно текущие индексы.

Скорее всего, такие полные дубликаты говорят о сбое при выгрузке данных. Необходимо сообщить об этом.

Еще одна проблема, о которой предупреждал менеджер проекта — дубликаты карточки Iskra Bar с неправильными названием бара - "Искра" и "Искр". На них заводились реальные заказы, поэтому необходимо вычислить их и переименовать на правильное название — Iskra Bar.

Объявим функцию `replace_restaurant()`, которая заменяет неправильное название, а потом проверяет успешность выполнения замены подсчетом (если замена прошла успешно — результат подсчета должен быть 0).

In [150]:
def replace_restaurant(data, name):
    data.restaurant.replace(name, 'Iskra Bar', inplace=True)
    return data[data['restaurant'] == name]['restaurant'].count()

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Я бы рекомендовал использовать метод для подсчета — `count()`, а не атрибут `shape`. 

Чтобы он считал только в нужной колонке restaurant (а не выводил все), можно использовать конструкцию `data[data['restaurant'] == name]['restaurant']`.

Также ты внутри функции явно пишешь `bo_data`, зачем тогда аргумент `data`?

Передаем функции названия 'Искра' и 'Искр'.

In [151]:
replace_restaurant(bo_data, 'Искра')

0

In [152]:
replace_restaurant(bo_data, 'Искр')

0

### Выводы

* Переиминовываем колонки: 'userID ' - 'user_ID', ' ресторан ' - 'restaurant', ' ordered ' - 'ordered_time', ' Столик для ' - 'table_for';
* Во время выгрузки данных произошел сбой, так как в датасете присутствуют полные дубликаты некоторых строк;
* В столбце 'restaurant' пропуски заменяем тестовой карточкой admin;
* Также заменяем карточки Iskra Bar c неверным названием ресторана (Искр, Искра) на верные.

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Здесь не столько вывод, сколько саммари, да: названия колонок, пропуски, карточка admin и тд 

## Распределение заказов в течение недели

Считаем, на какие кафе приходится больше всего заказов.

In [153]:
bo_data['restaurant'].value_counts()

Iskra Bar     287
Там Ем        200
Бар Сибирь    159
Северия       107
Крафтейнер     78
Коворкафе      66
Азия           62
admin          41
Name: restaurant, dtype: int64

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Не самая привычная запись (как минимум — для меня). Если что, обычно нужный столбец указывают в квадратных скобках `bo_data['restaurant']`.

Считаем, на какие дни приходится больше всего заказов.

In [154]:
bo_data['weekday'].value_counts()

Tuesday      251
Wednesday    242
Thursday     180
Monday       141
Friday       110
Saturday      52
Sunday        24
Name: weekday, dtype: int64

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Здесь нужно поправить описание шага. Под группировкой — имелся в виду метод `groupby()`, но вообще и со днями недели тоже можно с помощью `value_counts()` разобраться. Поправим

Объявляем функцию, которая возвращает количество заказов для ресторана в определенный день.

In [155]:
def number_of_orders(data, restaurant_name, day_of_the_week):
    return data[(data['restaurant'] == restaurant_name) & (data['weekday'] == day_of_the_week)]['restaurant'].count()

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Ага, круто, что логическое условие используешь, но про `shape` такой же коммент как и выше — попробуй `count()`. Можно сохранить лист, удовлетворяющий условию, а потом посчитать в нем количество ресторанов.

Передадим ей заведения Iskra Bar и Там Ем — два флагманских места, по которым просили результаты прежде всего — и понедельник, среду и пятницу в качестве дня недели.

In [156]:
iskra_mon = number_of_orders(bo_data, 'Iskra Bar', 'Monday')
print(iskra_mon)

48


In [157]:
iskra_wed = number_of_orders(bo_data, 'Iskra Bar', 'Wednesday')
print(iskra_wed)

53


In [158]:
iskra_fri = number_of_orders(bo_data, 'Iskra Bar', 'Friday')
print(iskra_fri)

29


In [159]:
tamem_mon = number_of_orders(bo_data, 'Там Ем', 'Monday')
print(tamem_mon)

25


In [160]:
tamem_wed = number_of_orders(bo_data, 'Там Ем', 'Wednesday')
print(tamem_wed)

52


In [161]:
tamem_fri = number_of_orders(bo_data, 'Там Ем', 'Friday')
print(tamem_fri)

24


Сведем результат в таблицу.

In [162]:
data = [
    ['Iskra Bar', iskra_mon, iskra_wed, iskra_fri],
    ['Там Ем', tamem_mon, tamem_wed, tamem_fri]
        ]

columns = ['Ресторан', 'Понедельник', 'Среда', 'Пятница']

table = pd.DataFrame(data=data, columns=columns)
table

Ресторан  Понедельник  Среда  Пятница
0  Iskra Bar           48     53       29
1     Там Ем           25     52       24

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Явно прописываешь `print()` — аутпут некрасивый получается. Если оставить только название переменной, то будет табличка. 

И еще закидывай присваивание результата выше (сейчас дублируешь), а всю ячейку оставь под сведение таблицы.

### Выводы

* Больше всего заказов приходится на Iskra Bar.
* Больше всего заказов приходится на вторник и среду (251 и 242 соответственно).
* Больше всего посетителей в Iskra Bar и Там Ем в среду.

## Распределение заказов в течение дня

Сравним количества заказов, сделанных в обеденное и вечернее время — с 12 до 16 и с 16 до 22. Для этого объявим функцию, которая принимает данные и 2 аргумента времени.

In [163]:
def timing(data, from_time, to_time):
    return data[(data['ordered_time'] >= from_time) & (data['ordered_time'] < to_time)]['ordered_time'].count()

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Столбцы через квадратные скобки и также `count()`.

Передадим промежуток с 12:00 и 16:00

In [164]:
timing(bo_data, '12:00', '16:00')

543

И с 16:00 до 22:00

In [165]:
timing(bo_data, '16:00', '22:00')

338

### Выводы

* Больше столиков бронируется в обеденное время.

## Типичное количество человек за столиком

Выясним, для какого количества человек обычно заказывают столик в разное время на примере Iskra Bar и Там Ем. Для этого напишем функцию, которая принимает данные, название ресторана, 2 времени и считает моду.

In [166]:
def mode_of_people(data, restaurant_name, from_time, to_time):
    return data[(data['ordered_time'] >= from_time) & (data['ordered_time'] < to_time) & (data['restaurant'] == restaurant_name)]['table_for'].mode()

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Столбцы через квадратные скобки и также `count()`.

Проверим функцию на двух ресторанах, а также дневном и вечернем промежутке времени.

In [167]:
mode_of_people(bo_data, 'Iskra Bar', '12:00', '16:00')

0    2
dtype: int64

In [168]:
mode_of_people(bo_data, 'Iskra Bar', '16:00', '22:00')

0    1
dtype: int64

In [169]:
mode_of_people(bo_data, 'Там Ем', '12:00', '16:00')

0    2
dtype: int64

In [170]:
mode_of_people(bo_data, 'Там Ем', '16:00', '22:00')

0    1
dtype: int64

### Выводы

* Днем обычно заказывают столик на двоих, а вечером – на одного.

<h3><font color="FFB800">Комментарий ассистента</font></h3>

Да, у больших компаний приложение не пользуется популярностью.

## Общие выводы

* Больше всего заказов приходится на Iskra Bar и Там Ем.
* Чаще столики бронируются на дневное время (c 12:00 до 16:00).
* Чаще всего в кафе приходят поодиночке или вдвоем.
* Наиболее загруженные дни в середине недели (вторник, среда и четверг).

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Cool

<h3><font color="FF0F00">Комментарий ассистента</font></h3>

Здесь лучше вернись к вопросам менеджера изначальным и ответь конкретно на них. 

Про третий вывод — Азия может иметь 60 посадочных мест и заполнять их на 100%, а Искра 500 и только половину. Если сравнивать так, то Азия популярнее :-))